### PR TITLE
Dataviz: Migrate to async DataSource APIs   

### LINKED ISSUE #127043: Phase 2.1 (@grafana/dataviz-squad): Migrate to async DataSource APIs
Part of #125083 — migrating internal usages of the legacy synchronous `DataSourceSrv` APIs to the new async APIs.

See the [parent issue](https://github.com/grafana/grafana/issues//125083) for migration patterns and the new async API reference.

## Files to migrate

- [ ] `public/app/features/actions/ConnectionPicker.tsx` — direct `getDataSourceSrv().getList()`; mapping reads `ds.meta.*` (needs the meta API)
- [ ] `public/app/features/dimensions/editors/FolderPickerTab.tsx` — indirect via `getDatasourceSrv().get()` (Catalog wrapper `datasource_srv.ts`)
- [ ] `public/app/plugins/panel/geomap/utils/utils.ts` — indirect via `getGrafanaDatasource()`; replace with `getDataSourceInstance('-- Grafana --')`

> Scope expanded from the original single file after an import-graph analysis of dataviz-owned code — see the findings comment below.

### PR DESCRIPTION
**Pull request by gelicia**
- `public/app/features/actions/ConnectionPicker.tsx` - There is a behavior change here, please see comments.
- `public/app/features/dimensions/editors/FolderPickerTab.tsx` 
- `public/app/plugins/panel/geomap/utils/utils.ts` - Not fixed directly, fixed `getGrafanaDatasource()` in `public/app/plugins/datasource/grafana/datasource.ts` 

Fixes: #127043

### CODE REVIEW

File: public/app/features/actions/ConnectionPicker.tsx [hunk 1]
```diff
@@ -74,13 +71,7 @@ export const ConnectionPicker = ({ actionType, datasourceUid, onChange, id }: Co
...
       onChange(DIRECT_OPTION_VALUE);
     } else {
-      const supportedDataSources = getSupportedDataSources();
-      const selectedDatasource = supportedDataSources.find((ds) => ds.uid === selectedValue);
-      if (selectedDatasource) {
-        onChange(selectedDatasource);
-      } else {
-        console.error('ConnectionPicker: Could not find datasource with UID:', selectedValue);
-      }
+      onChange(supportedDataSources.find((ds) => ds.uid === selectedValue)!);
```
Comment (gelicia) [inline public/app/features/actions/ConnectionPicker.tsx]: There is a behavior change here - previously, when you changed the connection, it would look up the datasource again. Now it loads the datasources when you load the picker, and doesn't recheck. This should only be an issue if you have this modal open, someone deletes a datasource, and then you select that datasource, but that seems like an edge case we shouldn't need to work around.

File: public/app/features/actions/ConnectionPicker.test.tsx [hunk 2]
```diff
@@ -133,23 +128,4 @@ describe('ConnectionPicker', () => {
 
     expect(onChange).toHaveBeenCalledWith('direct');
   });
-
-  it('logs an error and does not call onChange when the selected datasource cannot be found', async () => {
```
Comment (gelicia) [inline public/app/features/actions/ConnectionPicker.test.tsx:137]: This test was dependent on the old way of fetching datasources for this component and is no longer relevant. See https://github.com/grafana/grafana/pull/128155/changes#r3561093585

File: public/app/plugins/datasource/grafana/datasource.ts [hunk 3]
```diff
@@ -268,7 +263,8 @@ export class GrafanaDatasource extends DataSourceWithBackend<GrafanaQuery> {
 
 /** Get the GrafanaDatasource instance */
 export async function getGrafanaDatasource() {
-  return (await getDataSourceSrv().get('-- Grafana --')) as GrafanaDatasource;
+  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
+  return (await getDataSourceInstance('-- Grafana --')) as GrafanaDatasource;
```
Comment (gelicia) [inline public/app/plugins/datasource/grafana/datasource.ts]: This is changed for `public/app/plugins/panel/geomap/utils/utils.ts`

File: public/app/features/actions/ConnectionPicker.tsx [hunk 4]
```diff
@@ -45,21 +47,16 @@ export const ConnectionPicker = ({ actionType, datasourceUid, onChange, id }: Co
...
-        options.push({
-          label: ds.name,
-          value: ds.uid,
-          imgUrl: ds.meta.info.logos.small,
-        });
+    supportedDataSources.forEach((ds) => {
+      options.push({
+        label: ds.name,
+        value: ds.uid,
+        imgUrl: ds.meta.info.logos.small,
```
Comment (gelicia) [inline public/app/features/actions/ConnectionPicker.tsx:54]: There was a note in the issue about the logo not showing up, but it does fine? 
Image

**Review by wahajahmed010 — COMMENTED** [review]
## Review: Dataviz async DataSource API migration

**Overall:** Clean migration. The core change is straightforward — swapping getDataSourceSrv().getList() for the new getDataSourceInstanceList() from @grafana/runtime/unstable, and getDataSourceSrv().get() for getDataSourceInstance(). The diff is well-scoped (7 files, +74/-53).

### Good
- **Test coverage preserved and updated.** The ConnectionPicker test correctly mocks the new @grafana/runtime/unstable module instead of @grafana/runtime, and the async nature is handled with Promise.resolve.
- **Removed dead code.** The old error-logging test case for "selected datasource not found" was removed — the new code uses a non-null assertion (!) instead, which is a tradeoff worth noting (see below).
- **ESLint suppressions cleaned up.** Removing the FolderPickerTab suppressions and reducing the grafana datasource assertion count is a nice housekeeping bonus.

### Concerns

1. **Non-null assertion in ConnectionPicker** (supportedDataSources.find((ds) => ds.uid === selectedValue)!): The old code had a safety net — if the datasource wasn't found, it logged an error and didn't call onChange. The new code silently passes undefined (cast to the type via !). If the datasource list hasn't loaded yet (race condition on initial render), this could crash. Consider keeping the guard or at minimum a console.error.

2. **useEffect without cleanup:** The effect calls getDataSourceInstanceList(...).then(setSupportedDataSources) but doesn't cancel the promise on unmount. If the component unmounts before the promise resolves, setSupportedDataSources will fire on an unmounted component (React 18+ warns about this). A simple ignore flag or AbortController would help.

3. **useMemo dependency on supportedDataSources is correct** — good that it was changed from [] to [supportedDataSources]. The old code had an empty deps array which was a stale closure bug.

4. **FolderPickerTab noMargin change:** noMargin={false} is the default behavior, so this is a no-op. Was this intentional or leftover from debugging?

### Summary
Solid migration, functionally equivalent. The two main risks are the non-null assertion and the missing cleanup in the effect. Worth a quick fix before merging.

File: public/app/plugins/datasource/grafana/datasource.ts (hunk 3 above)
Comment (gtk-grafana) [inline public/app/plugins/datasource/grafana/datasource.ts]: We could avoid the assertion if we check the type at runtime, but I would understand if you'd rather stick with the assertion instead

```suggestion
  const ds = await getDataSourceInstance('-- Grafana --');
  if (!(ds instanceof GrafanaDatasource)) {
    throw new Error('Expected the "-- Grafana --" data source to be a GrafanaDatasource instance');
  }
  return ds;
```

File: public/app/features/actions/ConnectionPicker.tsx (hunk 4 above)
Comment (gtk-grafana) [reply public/app/features/actions/ConnectionPicker.tsx:54]: Yeah works for me too 🤷 

Image

File: public/app/features/actions/ConnectionPicker.tsx (hunk 1 above)
Comment (gtk-grafana) [inline public/app/features/actions/ConnectionPicker.tsx]: Nit: it would be nice to avoid the non-null assertion operator as it's pretty easy to overlook, I would argue being explicit that we won't call onChange if something isn't found is easier to read, but 🤷 
```suggestion
     const selected = supportedDataSources.find((ds) => ds.uid === selectedValue);
      if(selected){
        onChange(selected);
      }
```

File: public/app/plugins/datasource/grafana/datasource.ts (hunk 3 above)
Comment (gtk-grafana) [reply public/app/plugins/datasource/grafana/datasource.ts]: Oh but that would require updating the mock in the test:
```ts
// public/app/features/dimensions/editors/FolderPickerTab.test.tsx
jest.mock('app/plugins/datasource/grafana/datasource', () => ({
  ...jest.requireActual('app/plugins/datasource/grafana/datasource'),
  getGrafanaDatasource: () => Promise.resolve({ listFiles: mockListFiles }),
}));
```

**Review by gtk-grafana — DISMISSED** [review]
Some optional nits if we want to avoid the assertion and/or remove the non-null assertion operator but everything is working locally for me

**Comment by gelicia** [comment]
> 1. **Non-null assertion in ConnectionPicker** 

Fixed

> 2. **useEffect without cleanup:** 

I am not concerned with this scenario, as the call to the datasource is very small and the component is very unlikely to unmount before it is finished.

> 3. **useMemo dependency on supportedDataSources is correct** 

This does not appear to be a concern.

> 4. **FolderPickerTab noMargin change:** 

This is necessary to help resolve a linter rule https://github.com/grafana/grafana/blob/main/packages/grafana-eslint-rules/rules/no-restricted-syntax.cjs#L21

**Review by gtk-grafana — APPROVED** [review]
